### PR TITLE
[Disk Manager]: Add oauth2 jwt token exchange

### PR DESCRIPTION
**Pull request by jkuradobery**

### CODE REVIEW

File: cloud/disk_manager/internal/pkg/auth/credentials.go [hunk 1]
```diff
@@ -26,9 +28,27 @@ func NewCredentials(
 	if len(config.GetMetadataUrl()) == 0 {
 		return nil
 	}
-
-	return &credentialsWrapper{
-		impl: metadata.NewInstanceServiceAccount(
+	var impl Credentials
+	if config.GetServiceAccount() != nil {
+		_impl, err := credentials.NewOauth2TokenExchangeCredentials(
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/auth/credentials.go:33]:
```
var err error
...
impl, err = ...
```

File: cloud/disk_manager/internal/pkg/auth/credentials.go [hunk 2]
```diff
@@ -26,9 +28,27 @@ func NewCredentials(
...
+			credentials.WithTokenEndpoint(config.GetMetadataUrl()),
+			credentials.WithAudience(*config.ServiceAccount.Audience),
+			credentials.WithJWTSubjectToken(
+				credentials.WithSigningMethod(jwt.SigningMethodRS256),
+				credentials.WithKeyID(*config.ServiceAccount.KeyId),
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/auth/credentials.go:38]: config.GetServiceAccount().GetKeyId()

Comment (SvartMetal) [reply cloud/disk_manager/internal/pkg/auth/credentials.go:38]: and in other places

File: cloud/disk_manager/internal/pkg/auth/credentials.go [hunk 3]
```diff
@@ -26,9 +28,27 @@ func NewCredentials(
...
+				credentials.WithRSAPrivateKeyPEMFile(config.GetCertFile()),
+				credentials.WithIssuer(*config.ServiceAccount.Id),
+				credentials.WithSubject(*config.ServiceAccount.Id),
+				credentials.WithAudience(*config.ServiceAccount.Audience),
+			),
+		)
+		if err != nil {
+			logging.Error(ctx, "failed to create token credential, error: %s", err)
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/auth/credentials.go:46]: failed to create token credentials

Comment (jkuradobery) [reply cloud/disk_manager/internal/pkg/auth/credentials.go:46]: ./credentials.go:50:4: github.com/ydb-platform/nbs/cloud/tasks/logging.Error does not support error-wrapping directive %w
I would stick with %s

File: cloud/disk_manager/internal/pkg/auth/credentials.go [hunk 4]
```diff
@@ -26,9 +28,30 @@ func NewCredentials(
...
+			credentials.WithTokenEndpoint(config.GetMetadataUrl()),
+			credentials.WithAudience(config.GetServiceAccount().GetAudience()),
+			credentials.WithJWTSubjectToken(
+				credentials.WithSigningMethod(jwt.SigningMethodRS256),
+				credentials.WithKeyID(config.GetServiceAccount().GetKeyId()),
+				credentials.WithRSAPrivateKeyPEMFile(config.GetCertFile()),
+				credentials.WithIssuer(config.GetServiceAccount().GetId()),
+				credentials.WithSubject(config.GetServiceAccount().GetId()),
+				credentials.WithAudience(
+					config.GetServiceAccount().GetAudience()),
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/auth/credentials.go:46]:
```
				credentials.WithAudience(
					config.GetServiceAccount().GetAudience(),
                                 ),
```

Comment (SvartMetal) [reply cloud/disk_manager/internal/pkg/auth/credentials.go:46]: Либо пусть будет на одной строке

File: cloud/disk_manager/internal/pkg/auth/credentials.go [hunk 5]
```diff
@@ -26,9 +28,30 @@ func NewCredentials(
...
+			),
+		)
+		if err != nil {
+			logging.Error(ctx, "failed to create token credentials: %s", err)
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/auth/credentials.go:50]: %s -> %w

File: cloud/disk_manager/internal/pkg/auth/credentials.go [hunk 6]
```diff
@@ -26,9 +28,30 @@ func NewCredentials(
...
+			return nil
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/auth/credentials.go:51]: Why?
This function should return an error

Comment (jkuradobery) [reply cloud/disk_manager/internal/pkg/auth/credentials.go:51]: Let us make this function return an error in another PR.

File: cloud/disk_manager/internal/pkg/auth/credentials.go (hunk 5 above)
Comment (SvartMetal) [reply cloud/disk_manager/internal/pkg/auth/credentials.go:50]: %v

**Review by SvartMetal — APPROVED** [review]

**Review by qkrorlqr — APPROVED** [review]

**Review by SvartMetal — APPROVED** [review]

**Pull request merged**